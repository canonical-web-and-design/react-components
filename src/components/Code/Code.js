import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useRef } from "react";

/**
 * @deprecated Code component is deprecated. Use CodeSnippet component or inline `<code>` instead.
 */
const Code = ({
  children,
  className,
  inline,
  copyable,
  numbered,
  ...props
}) => {
  const input = useRef(null);
  if (inline) {
    return (
      <code className={className} {...props}>
        {children}
      </code>
    );
  } else if (copyable) {
    const handleClick = (evt) => {
      input.current.focus();
      input.current.select();
      try {
        document.execCommand("copy");
      } catch (err) {
        console.error("Copy was unsuccessful");
      }
    };
    return (
      <div className={classNames(className, "p-code-copyable")} {...props}>
        <input
          className="p-code-copyable__input"
          readOnly="readonly"
          ref={input}
          value={children}
        />
        <button className="p-code-copyable__action" onClick={handleClick}>
          Copy to clipboard
        </button>
      </div>
    );
  } else if (numbered) {
    const lines = children.split(/\r?\n/);
    const content = lines.map((line, i) => (
      <span className="p-code-numbered__line" key={i}>
        {line}
      </span>
    ));
    return (
      <pre className={classNames(className, "p-code-numbered")} {...props}>
        <code>{content}</code>
      </pre>
    );
  } else {
    return (
      <pre className={className} {...props}>
        <code>{children}</code>
      </pre>
    );
  }
};

Code.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  inline: PropTypes.bool,
  copyable: PropTypes.bool,
  numbered: PropTypes.bool,
};

export default Code;
