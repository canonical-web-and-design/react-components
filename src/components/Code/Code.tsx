import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useRef } from "react";
import type { HTMLProps } from "react";

type Props = {
  children: string;
  className?: string;
  inline?: boolean;
  copyable?: boolean;
  numbered?: boolean;
} & HTMLProps<HTMLElement>;

const Code = ({
  children,
  className,
  inline,
  copyable,
  numbered,
  ...props
}: Props): JSX.Element => {
  const input = useRef(null);
  if (inline) {
    return (
      <code className={className} {...props}>
        {children}
      </code>
    );
  } else if (copyable) {
    const handleClick = () => {
      input.current.focus();
      input.current.select();
      try {
        document.execCommand("copy");
      } catch (err) {
        console.error("Copy was unsuccessful");
      }
    };
    return (
      <div className={classNames(className, "p-code-copyable")} {...props}>
        <input
          className="p-code-copyable__input"
          ref={input}
          value={children}
          readOnly
        />
        <button className="p-code-copyable__action" onClick={handleClick}>
          Copy to clipboard
        </button>
      </div>
    );
  } else if (numbered) {
    const lines = children.split(/\r?\n/);
    const content = lines.map((line, i) => (
      <span className="p-code-numbered__line" key={i}>
        {line}
      </span>
    ));
    return (
      <pre className={classNames(className, "p-code-numbered")} {...props}>
        <code>{content}</code>
      </pre>
    );
  } else {
    return (
      <pre className={className} {...props}>
        <code>{children}</code>
      </pre>
    );
  }
};

Code.propTypes = {
  children: PropTypes.string,
  className: PropTypes.string,
  inline: PropTypes.bool,
  copyable: PropTypes.bool,
  numbered: PropTypes.bool,
};

export default Code;
