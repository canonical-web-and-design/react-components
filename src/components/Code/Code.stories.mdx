import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Code from "./Code";

<Meta title="Code" component={Code} />

### Code

This is a [React](https://reactjs.org/) component for the Vanilla [Code](https://docs.vanillaframework.io/base/code/).

Vanilla gives you multiple ways to display code using the standard HTML elements.

### Props

<ArgsTable of={Code} />

### Inline

<Canvas>
  <Story name="Inline">
    <p>
      The quick brown <Code inline>fox jumps</Code> over the lazy dog
    </p>
  </Story>
</Canvas>

### Block

<Canvas>
  <Story name="Block">
    <Code>Mode: all Settings: maas_url=http://192.168.122.1:5240/MAAS</Code>
  </Story>
</Canvas>

### Copyable

<Canvas>
  <Story name="Copyable">
    <Code copyable>sudo apt-get update</Code>
  </Story>
</Canvas>

### Numbered

<Canvas>
  <Story name="Numbered">
    <Code numbered>
      #!/bin/bash set -eu . $CONJURE_UP_SPELLSDIR/sdk/common.sh if [[
      "$JUJU_PROVIDERTYPE" == "lxd" ]]; then debug "Running pre-deploy for
      $CONJURE_UP_SPELL" sed "s/##MODEL##/$JUJU_MODEL/"
      $(scriptPath)/lxd-profile.yaml | lxc profile edit "juju-$JUJU_MODEL" ||
      exposeResult "Failed to set profile" $? "false" fi exposeResult
      "Successful pre-deploy." 0 "true"
    </Code>
  </Story>
</Canvas>
