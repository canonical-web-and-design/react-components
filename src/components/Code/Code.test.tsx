import { shallow } from "enzyme";
import React from "react";

import Code from "./Code";

describe("Code ", () => {
  it("can render", () => {
    const wrapper = shallow(<Code>Test content</Code>);
    expect(wrapper).toMatchSnapshot();
  });

  it("can render inline", () => {
    const wrapper = shallow(<Code inline>Test content</Code>);
    expect(wrapper).toMatchSnapshot();
  });

  it("can render as copyable", () => {
    const wrapper = shallow(<Code copyable>Test content</Code>);
    expect(wrapper).toMatchSnapshot();
  });

  it("can render with line numbers", () => {
    const wrapper = shallow(<Code numbered>Test content</Code>);
    expect(wrapper).toMatchSnapshot();
  });
});
