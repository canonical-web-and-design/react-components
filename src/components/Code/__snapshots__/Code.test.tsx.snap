// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code  can render 1`] = `
<pre>
  <code>
    Test content
  </code>
</pre>
`;

exports[`Code  can render as copyable 1`] = `
<div
  className="p-code-copyable"
>
  <input
    className="p-code-copyable__input"
    readOnly={true}
    value="Test content"
  />
  <button
    className="p-code-copyable__action"
    onClick={[Function]}
  >
    Copy to clipboard
  </button>
</div>
`;

exports[`Code  can render inline 1`] = `
<code>
  Test content
</code>
`;

exports[`Code  can render with line numbers 1`] = `
<pre
  className="p-code-numbered"
>
  <code>
    <span
      className="p-code-numbered__line"
      key="0"
    >
      Test content
    </span>
  </code>
</pre>
`;
